File: backend/govinda-masterdata/src/main/java/net/voytrex/govinda/masterdata/domain/model/Person.java
/*
 * Govinda ERP - Person Entity
 * Copyright 2024 Voytrex
 * SPDX-License-Identifier: Apache-2.0
 */

package net.voytrex.govinda.masterdata.domain.model;

import jakarta.persistence.AttributeOverride;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Embedded;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import jakarta.persistence.Version;
import java.time.Instant;
import java.time.LocalDate;
import java.time.Period;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.UUID;
import net.voytrex.govinda.common.domain.model.AddressType;
import net.voytrex.govinda.common.domain.model.AgeGroup;
import net.voytrex.govinda.common.domain.model.AhvNumber;
import net.voytrex.govinda.common.domain.model.Gender;
import net.voytrex.govinda.common.domain.model.Historized;
import net.voytrex.govinda.common.domain.model.Language;
import net.voytrex.govinda.common.domain.model.MaritalStatus;
import net.voytrex.govinda.common.domain.model.MutationType;
import net.voytrex.govinda.common.domain.model.PersonStatus;

/**
 * Person entity representing an insured individual.
 *
 * Supports history tracking for name and marital status changes.
 */
@Entity
@Table(name = "person")
public class Person implements Historized<PersonHistoryEntry> {
    @Id
    @Column(name = "id", updatable = false, nullable = false)
    private UUID id = UUID.randomUUID();

    @Column(name = "tenant_id", updatable = false, nullable = false)
    private UUID tenantId;

    @Embedded
    @AttributeOverride(name = "value", column = @Column(name = "ahv_nr", length = 16, nullable = false))
    private AhvNumber ahvNr;

    @Column(name = "last_name", length = 100, nullable = false)
    private String lastName;

    @Column(name = "first_name", length = 100, nullable = false)
    private String firstName;

    @Column(name = "date_of_birth", nullable = false)
    private LocalDate dateOfBirth;

    @Enumerated(EnumType.STRING)
    @Column(name = "gender", length = 10, nullable = false)
    private Gender gender;

    @Enumerated(EnumType.STRING)
    @Column(name = "marital_status", length = 20)
    private MaritalStatus maritalStatus;

    @Column(name = "nationality", length = 3)
    private String nationality = "CHE";

    @Enumerated(EnumType.STRING)
    @Column(name = "preferred_language", length = 2)
    private Language preferredLanguage = Language.DE;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", length = 20, nullable = false)
    private PersonStatus status = PersonStatus.ACTIVE;

    @Column(name = "created_at", updatable = false, nullable = false)
    private Instant createdAt = Instant.now();

    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt = Instant.now();

    @Version
    @Column(name = "version", nullable = false)
    private long version = 0L;

    @OneToMany(mappedBy = "person", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<Address> addresses = new ArrayList<>();

    protected Person() {
    }

    public Person(
        UUID tenantId,
        AhvNumber ahvNr,
        String lastName,
        String firstName,
        LocalDate dateOfBirth,
        Gender gender,
        MaritalStatus maritalStatus,
        String nationality,
        Language preferredLanguage
    ) {
        if (lastName == null || lastName.isBlank()) {
            throw new IllegalArgumentException("Last name must not be blank");
        }
        if (firstName == null || firstName.isBlank()) {
            throw new IllegalArgumentException("First name must not be blank");
        }
        if (dateOfBirth.isAfter(LocalDate.now())) {
            throw new IllegalArgumentException("Date of birth cannot be in the future");
        }
        this.tenantId = tenantId;
        this.ahvNr = ahvNr;
        this.lastName = lastName;
        this.firstName = firstName;
        this.dateOfBirth = dateOfBirth;
        this.gender = gender;
        this.maritalStatus = maritalStatus;
        if (nationality != null) {
            this.nationality = nationality;
        }
        if (preferredLanguage != null) {
            this.preferredLanguage = preferredLanguage;
        }
    }

    @Override
    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public UUID getTenantId() {
        return tenantId;
    }

    public void setTenantId(UUID tenantId) {
        this.tenantId = tenantId;
    }

    public AhvNumber getAhvNr() {
        return ahvNr;
    }

    public void setAhvNr(AhvNumber ahvNr) {
        this.ahvNr = ahvNr;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        if (lastName == null || lastName.isBlank()) {
            throw new IllegalArgumentException("Last name must not be blank");
        }
        this.lastName = lastName;
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        if (firstName == null || firstName.isBlank()) {
            throw new IllegalArgumentException("First name must not be blank");
        }
        this.firstName = firstName;
    }

    public LocalDate getDateOfBirth() {
        return dateOfBirth;
    }

    public void setDateOfBirth(LocalDate dateOfBirth) {
        this.dateOfBirth = dateOfBirth;
    }

    public Gender getGender() {
        return gender;
    }

    public void setGender(Gender gender) {
        this.gender = gender;
    }

    public MaritalStatus getMaritalStatus() {
        return maritalStatus;
    }

    public void setMaritalStatus(MaritalStatus maritalStatus) {
        this.maritalStatus = maritalStatus;
    }

    public String getNationality() {
        return nationality;
    }

    public void setNationality(String nationality) {
        this.nationality = nationality;
    }

    public Language getPreferredLanguage() {
        return preferredLanguage;
    }

    public void setPreferredLanguage(Language preferredLanguage) {
        this.preferredLanguage = preferredLanguage;
    }

    public PersonStatus getStatus() {
        return status;
    }

    public void setStatus(PersonStatus status) {
        this.status = status;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(Instant createdAt) {
        this.createdAt = createdAt;
    }

    public Instant getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(Instant updatedAt) {
        this.updatedAt = updatedAt;
    }

    public long getVersion() {
        return version;
    }

    public void setVersion(long version) {
        this.version = version;
    }

    public List<Address> getAddresses() {
        return List.copyOf(addresses);
    }

    public void setAddresses(List<Address> addresses) {
        this.addresses = addresses;
    }

    /**
     * Returns the full name (first name + last name).
     */
    public String fullName() {
        return firstName + " " + lastName;
    }

    /**
     * Calculates the age at a given date.
     */
    public int ageAt(LocalDate date) {
        return Period.between(dateOfBirth, date).getYears();
    }

    /**
     * Determines the age group for premium calculation at a given date.
     */
    public AgeGroup ageGroupAt(LocalDate date) {
        return AgeGroup.forAge(ageAt(date));
    }

    /**
     * Returns the current (active) main address.
     */
    public Address currentAddress() {
        return addresses.stream()
            .filter(address -> address.getAddressType() == AddressType.MAIN && address.isCurrent())
            .findFirst()
            .orElse(null);
    }

    /**
     * Returns the address valid at a specific date.
     */
    public Address addressAt(LocalDate date) {
        return addresses.stream()
            .filter(address -> address.getAddressType() == AddressType.MAIN && address.isValidOn(date))
            .findFirst()
            .orElse(null);
    }

    /**
     * Adds a new address, closing the previous one if exists.
     */
    public void addAddress(Address address, LocalDate closeExistingOn) {
        if (closeExistingOn != null) {
            Address current = currentAddress();
            if (current != null) {
                current.close(closeExistingOn);
            }
        }
        addresses.add(address);
        updatedAt = Instant.now();
    }

    /**
     * Changes the name (e.g., due to marriage) and creates a history entry.
     */
    public PersonHistoryEntry changeName(
        String newLastName,
        String newFirstName,
        String reason,
        LocalDate effectiveDate,
        UUID changedBy
    ) {
        if (newLastName == null || newLastName.isBlank()) {
            throw new IllegalArgumentException("New last name must not be blank");
        }
        if (newFirstName == null || newFirstName.isBlank()) {
            throw new IllegalArgumentException("New first name must not be blank");
        }

        PersonHistoryEntry historyEntry = createHistoryEntry(
            MutationType.UPDATE,
            reason,
            changedBy
        );
        historyEntry.setValidTo(effectiveDate.minusDays(1));

        this.lastName = newLastName;
        this.firstName = newFirstName;
        this.updatedAt = Instant.now();

        return historyEntry;
    }

    /**
     * Changes the marital status and creates a history entry.
     */
    public PersonHistoryEntry changeMaritalStatus(
        MaritalStatus newStatus,
        String reason,
        LocalDate effectiveDate,
        UUID changedBy
    ) {
        PersonHistoryEntry historyEntry = createHistoryEntry(
            MutationType.UPDATE,
            reason,
            changedBy
        );
        historyEntry.setValidTo(effectiveDate.minusDays(1));

        this.maritalStatus = newStatus;
        this.updatedAt = Instant.now();

        return historyEntry;
    }

    @Override
    public PersonHistoryEntry createHistoryEntry(MutationType mutationType, String reason, UUID changedBy) {
        return new PersonHistoryEntry(
            this.id,
            this.lastName,
            this.firstName,
            this.maritalStatus,
            LocalDate.now(),
            null,
            mutationType,
            reason,
            changedBy
        );
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (other == null || getClass() != other.getClass()) {
            return false;
        }
        Person person = (Person) other;
        return Objects.equals(id, person.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    @Override
    public String toString() {
        return "Person(id=" + id + ", ahvNr=" + ahvNr + ", name='" + firstName + " " + lastName + "')";
    }
}

File: backend/govinda-masterdata/src/main/kotlin/net/voytrex/govinda/masterdata/domain/model/Person.kt
/*
 * Govinda ERP - Person Entity
 * Copyright 2024 Voytrex
 * SPDX-License-Identifier: Apache-2.0
 */

package net.voytrex.govinda.masterdata.domain.model

import jakarta.persistence.*
import net.voytrex.govinda.common.domain.model.*
import java.time.Instant
import java.time.LocalDate
import java.time.Period
import java.util.UUID

/**
 * Person entity representing an insured individual.
 *
 * This is a core entity in the system, representing natural persons
 * who can be insured under health insurance policies.
 *
 * Supports history tracking for name and marital status changes.
 */
@Entity
@Table(name = "person")
class Person(
    @Id
    @Column(name = "id", updatable = false, nullable = false)
    override val id: UUID = UUID.randomUUID(),

    @Column(name = "tenant_id", updatable = false, nullable = false)
    val tenantId: UUID,

    @Embedded
    @AttributeOverride(name = "value", column = Column(name = "ahv_nr", length = 16, nullable = false))
    val ahvNr: AhvNumber,

    @Column(name = "last_name", length = 100, nullable = false)
    var lastName: String,

    @Column(name = "first_name", length = 100, nullable = false)
    var firstName: String,

    @Column(name = "date_of_birth", nullable = false)
    val dateOfBirth: LocalDate,

    @Enumerated(EnumType.STRING)
    @Column(name = "gender", length = 10, nullable = false)
    val gender: Gender,

    @Enumerated(EnumType.STRING)
    @Column(name = "marital_status", length = 20)
    var maritalStatus: MaritalStatus? = null,

    @Column(name = "nationality", length = 3)
    var nationality: String? = "CHE",

    @Enumerated(EnumType.STRING)
    @Column(name = "preferred_language", length = 2)
    var preferredLanguage: Language = Language.DE,

    @Enumerated(EnumType.STRING)
    @Column(name = "status", length = 20, nullable = false)
    var status: PersonStatus = PersonStatus.ACTIVE,

    @Column(name = "created_at", updatable = false, nullable = false)
    val createdAt: Instant = Instant.now(),

    @Column(name = "updated_at", nullable = false)
    var updatedAt: Instant = Instant.now(),

    @Version
    @Column(name = "version", nullable = false)
    var version: Long = 0

) : Historized<PersonHistoryEntry> {

    @OneToMany(mappedBy = "person", cascade = [CascadeType.ALL], orphanRemoval = true, fetch = FetchType.LAZY)
    private val _addresses: MutableList<Address> = mutableListOf()

    val addresses: List<Address>
        get() = _addresses.toList()

    init {
        require(lastName.isNotBlank()) { "Last name must not be blank" }
        require(firstName.isNotBlank()) { "First name must not be blank" }
        require(!dateOfBirth.isAfter(LocalDate.now())) { "Date of birth cannot be in the future" }
    }

    // ═══════════════════════════════════════════════════════════════
    // Business Methods
    // ═══════════════════════════════════════════════════════════════

    /**
     * Returns the full name (first name + last name).
     */
    fun fullName(): String = "$firstName $lastName"

    /**
     * Calculates the age at a given date.
     */
    fun ageAt(date: LocalDate): Int = Period.between(dateOfBirth, date).years

    /**
     * Determines the age group for premium calculation at a given date.
     */
    fun ageGroupAt(date: LocalDate): AgeGroup = AgeGroup.forAge(ageAt(date))

    /**
     * Returns the current (active) main address.
     */
    fun currentAddress(): Address? =
        _addresses.find { it.addressType == AddressType.MAIN && it.isCurrent() }

    /**
     * Returns the address valid at a specific date.
     */
    fun addressAt(date: LocalDate): Address? =
        _addresses.find { it.addressType == AddressType.MAIN && it.isValidOn(date) }

    /**
     * Adds a new address, closing the previous one if exists.
     */
    fun addAddress(address: Address, closeExistingOn: LocalDate? = null) {
        if (closeExistingOn != null) {
            currentAddress()?.close(closeExistingOn)
        }
        _addresses.add(address)
        updatedAt = Instant.now()
    }

    // ═══════════════════════════════════════════════════════════════
    // History-Creating Mutations
    // ═══════════════════════════════════════════════════════════════

    /**
     * Changes the name (e.g., due to marriage) and creates a history entry.
     *
     * @param newLastName The new last name
     * @param newFirstName The new first name (defaults to current)
     * @param reason The reason for the change
     * @param effectiveDate When the change takes effect
     * @param changedBy The user making the change
     * @return History entry with the old values
     */
    fun changeName(
        newLastName: String,
        newFirstName: String = this.firstName,
        reason: String,
        effectiveDate: LocalDate,
        changedBy: UUID
    ): PersonHistoryEntry {
        require(newLastName.isNotBlank()) { "New last name must not be blank" }
        require(newFirstName.isNotBlank()) { "New first name must not be blank" }

        val historyEntry = createHistoryEntry(
            mutationType = MutationType.UPDATE,
            reason = reason,
            changedBy = changedBy
        ).copy(validTo = effectiveDate.minusDays(1))

        this.lastName = newLastName
        this.firstName = newFirstName
        this.updatedAt = Instant.now()

        return historyEntry
    }

    /**
     * Changes the marital status and creates a history entry.
     */
    fun changeMaritalStatus(
        newStatus: MaritalStatus,
        reason: String,
        effectiveDate: LocalDate,
        changedBy: UUID
    ): PersonHistoryEntry {
        val historyEntry = createHistoryEntry(
            mutationType = MutationType.UPDATE,
            reason = reason,
            changedBy = changedBy
        ).copy(validTo = effectiveDate.minusDays(1))

        this.maritalStatus = newStatus
        this.updatedAt = Instant.now()

        return historyEntry
    }

    override fun createHistoryEntry(
        mutationType: MutationType,
        reason: String?,
        changedBy: UUID
    ) = PersonHistoryEntry(
        personId = this.id,
        lastName = this.lastName,
        firstName = this.firstName,
        maritalStatus = this.maritalStatus,
        validFrom = LocalDate.now(),
        validTo = null,
        mutationType = mutationType,
        mutationReason = reason,
        changedBy = changedBy
    )

    // ═══════════════════════════════════════════════════════════════
    // Equality
    // ═══════════════════════════════════════════════════════════════

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || this::class != other::class) return false
        other as Person
        return id == other.id
    }

    override fun hashCode(): Int = id.hashCode()

    override fun toString(): String = "Person(id=$id, ahvNr=$ahvNr, name='$firstName $lastName')"
}

/**
 * History entry for Person changes.
 */
@Entity
@Table(name = "person_history")
data class PersonHistoryEntry(
    @Id
    @Column(name = "history_id")
    override val historyId: UUID = UUID.randomUUID(),

    @Column(name = "person_id", nullable = false)
    val personId: UUID,

    @Column(name = "last_name", length = 100, nullable = false)
    val lastName: String,

    @Column(name = "first_name", length = 100, nullable = false)
    val firstName: String,

    @Enumerated(EnumType.STRING)
    @Column(name = "marital_status", length = 20)
    val maritalStatus: MaritalStatus?,

    @Column(name = "valid_from", nullable = false)
    override val validFrom: LocalDate,

    @Column(name = "valid_to")
    override val validTo: LocalDate?,

    @Column(name = "recorded_at", nullable = false)
    override val recordedAt: Instant = Instant.now(),

    @Column(name = "superseded_at")
    override val supersededAt: Instant? = null,

    @Enumerated(EnumType.STRING)
    @Column(name = "mutation_type", length = 20, nullable = false)
    override val mutationType: MutationType,

    @Column(name = "mutation_reason", length = 500)
    override val mutationReason: String?,

    @Column(name = "changed_by", nullable = false)
    override val changedBy: UUID

) : HistoryEntry(
    historyId, validFrom, validTo, recordedAt,
    supersededAt, mutationType, mutationReason, changedBy
)
